`gh agent task create`: Allow interactively editing task description file and add confirmation prompt

BagToad opened the pull request:
This pull request refactors and improves the handling of task descriptions in the `agent-task create` command, especially around the use of the `--from-file` (`-F`) flag, and enhances related tests for clarity and coverage. The changes separate file loading logic from argument parsing, improve error handling and user prompts, and modernize the test suite.

![cap](https://github.com/user-attachments/assets/69fe5d0d-f299-4390-a5dc-36be17ccd207)

BagToad committed:
Refactor agent-task create to improve file input handling

Moves file reading for the problem statement from argument parsing to execution, storing the file path in CreateOptions. Updates error handling and user prompts to better distinguish between interactive and non-interactive modes. Refactors and expands tests to match the new logic and improve coverage for file and prompt scenarios.

BagToad committed:
Update command examples for agent-task create

Improves the help text for the 'gh agent-task create' command by clarifying the editor usage example and adding an example for using a file as a template.

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -110,20 +104,37 @@ func createRun(opts *CreateOptions) error {
 		if err != nil {
 			return err
 		}
-
-		trimmed := strings.TrimSpace(desc)
-		if trimmed == "" {
-			return cmdutil.FlagErrorf("a task description is required")
+		if !confirm {
+			return cmdutil.SilentError
 		}
-		opts.ProblemStatement = trimmed
 	}
```
Question; do we want to always ask for confirmation? even when the problem statement is provided as an argument?

BagToad replied:
I think it aligns more with `pr create` and other commands to _not prompt_ when it's given as an arg. I think this sort of ergonomics makes sense to me, but maybe it's just what I'm used to? Could be something to add later.

babakks replied:
Just to make myself clear, yeah I think we shouldn't prompt if there's an arg provided. But current PR does prompt for confirmation regardless.

BagToad replied:
I mean, `gh pr create` accepts the body as a flag arg, so it's not quite the same, but I'm loosely aligning with that.

BagToad replied:
> Just to make myself clear, yeah I think we shouldn't prompt if there's an arg provided. But current PR does prompt for confirmation regardless.

Sorry this reply didn't load before I responded. I see what you're saying now 🤔

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -69,29 +67,41 @@ func TestNewCmdCreate(t *testing.T) {
+				ProblemStatement:     "",
+				ProblemStatementFile: "-",
+			},
 		},
```
Since we no longer perform file IO in flag parsing, I think these two test cases are not needed, esp. since the actual file is not there to have a real test against an empty file/stdin.

BagToad replied:
Thanks - fixed in 872cf495c24be303183a182d64b2b22e15a0c735. Also added a test for basebranch.

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -21,46 +20,45 @@ import (
 	tests := []struct {
 		name     string
 		args     string
 		stdin    string
```
The `stdin` field is no longer needed.

BagToad replied:
Fixed in 666f5574e4f8074063370c54c5e0cf2cf7d2ac3c

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -157,64 +171,149 @@ func Test_createRun(t *testing.T) {
+		name       string
+		isTTY      bool
+		capiStubs  func(*testing.T, *capi.CapiClientMock)
+		opts       *CreateOptions // input options (IO & BackOff set later)
```
Thanks for this! 🙏

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -157,64 +171,149 @@ func Test_createRun(t *testing.T) {
+					MarkdownEditorFunc: func(prompt, defaultValue string, blankAllowed bool) (string, error) {
+						require.Equal(t, "Enter the task description", prompt)
+						require.Equal(t, "task description from file", defaultValue)
+						return prompt + "+From editor", nil
```
Question; why we're including the prompt in the fake editor return (i.e `prompt + "..."`)?

BagToad replied:
Probably just I was tired and wanted to communicate that the prompt can be changed from what is in the file - the string concatenation is not needed, so fixed with directly returning an altered value in f22dc9271b19c31f41b43bb400eaa06a3b69ac45

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -157,64 +171,149 @@ func Test_createRun(t *testing.T) {
+			wantErr:    "SilentError",
+			wantStdErr: "a task description is required.\n",
+		},
+		{
```
I think we also need another test case for a non-empty arg.

```go
		{
			name: "problem statement loaded from arg non-interactively doesn't prompt or return error",
			opts: &CreateOptions{
				BaseRepo:             func() (ghrepo.Interface, error) { return ghrepo.New("OWNER", "REPO"), nil },
				ProblemStatement:     "task description",
			},
			capiStubs: func(t *testing.T, m *capi.CapiClientMock) {
				m.CreateJobFunc = func(ctx context.Context, owner, repo, problemStatement, baseBranch string) (*capi.Job, error) {
					require.Equal(t, "OWNER", owner)
					require.Equal(t, "REPO", repo)
					require.Equal(t, "task description", problemStatement)
					return &createdJobSuccessWithPR, nil
				}
			},
			wantStdout: "https://github.com/OWNER/REPO/pull/42/agent-sessions/sess1\n",
		},
```

BagToad replied:
Good call. Added in 1dbb694790ba59c12958a254b78e8030adb755a2

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -157,64 +171,149 @@ func Test_createRun(t *testing.T) {
+			},
+			isTTY:      true,
+			wantErr:    "SilentError",
+			wantStdErr: "",
```
What we want here to is to make sure error is the right type. I suggest adding a `wantErrIs` of type `error`, and assert it using `require.ErrorIs` function. We have already done that in `capi_test.go`:

https://github.com/cli/cli/blob/deee0c61eda2a897dc25faa21a2828611d5748e9/pkg/cmd/agent-task/capi/sessions_test.go#L1488-L1492

https://github.com/cli/cli/blob/deee0c61eda2a897dc25faa21a2828611d5748e9/pkg/cmd/agent-task/capi/sessions_test.go#L1729-L1737

BagToad replied:
Good idea. Fixed in 3831380d131bd5fde42186a051f10e584e341e84

babakks reviewed (approved):
LGTM! Just a couple of questions/suggestions.

BagToad committed:
Refactor create command tests and add base branch case

Removed redundant test cases for file input and stdin in TestNewCmdCreate, and added a new test to verify that the base branch argument sets the BaseBranch field correctly.

BagToad committed:
Remove unused stdin field from create command tests

The 'stdin' field and related code were removed from TestNewCmdCreate as they were no longer used. This simplifies the test structure and eliminates unnecessary code.

BagToad committed:
Update test to use edited task description

Modified the test in create_test.go to return and expect 'edited task description' instead of concatenating the prompt string. This clarifies the test's intent and expected behavior for the MarkdownEditorFunc and CreateJobFunc.

BagToad committed:
Add test for non-interactive problem statement input

Adds a test case to ensure that when a problem statement is provided as an argument non-interactively, the command does not prompt or return an error, and the correct job is created.

BagToad committed:
Add error type assertion to createRun tests

Introduces a new 'wantErrIs' field to test cases in Test_createRun to assert specific error types using require.ErrorIs. This enhances test coverage by verifying not only error messages but also error types.
